src/inference/graph_inference.cpp

```diff
@@ -341,6 +342,24 @@ class GraphNodeAssignment : public Nice2Assignment {
     return label_set_->GetLabelName(label_id);
   }
 
+  // Returns conditional probability of a label conditionated on its neighbouring labels
+  double GetNodeConditionalProbabilityGivenNeighbours(const GraphInference& fweights, int node, const double normalization, int label) const {
+    return exp(GetNodeScoreGivenAssignmentToANode(fweights, node, node, label)) / normalization;
+  }
+
+  // Returns the normalization related to a certain node
+  double GetNodeNormalizationConstantGivenNeighbours(const GraphInference& fweights, int node, std::vector<int>* candidates) const {
```

Comment:
In the normalisation, the first thing that I do is to add to the candidates vector the current  assignment of the node to which I am computing the estimated normalisation. As I said before now I see why this is better inside PLLearn. It would be more readable.